File: chapters/03-domain-driven-design-and-bounded-contexts/java/OrderAggregate.java
// Scenario: checkout order — lines cannot change after submit
// Demonstrates: Aggregate root enforcing invariants
// Trade-off: small aggregate vs loading entire order graph in one transaction

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

enum OrderStatus { DRAFT, SUBMITTED }

record LineItem(String sku, int qty) {}

final class Order {
    private final String orderId;
    private OrderStatus status = OrderStatus.DRAFT;
    private final List<LineItem> lines = new ArrayList<>();

    Order(String orderId) { this.orderId = Objects.requireNonNull(orderId); }

    void addLine(LineItem item) {
        if (status != OrderStatus.DRAFT) throw new IllegalStateException("cannot modify submitted order");
        lines.add(item);
    }

    void submit() {
        if (lines.isEmpty()) throw new IllegalStateException("empty order");
        status = OrderStatus.SUBMITTED;
    }

    String orderId() { return orderId; }
    OrderStatus status() { return status; }
    List<LineItem> lines() { return Collections.unmodifiableList(lines); }
}

File: docs/chapters/03-domain-driven-design-and-bounded-contexts/go/order_aggregate.go
// Scenario: checkout order — lines cannot change after submit
// Demonstrates: Aggregate root enforcing invariants
// Trade-off: exported fields vs methods; keep invariants inside one type

package main

import "errors"

type OrderStatus string

const (
	StatusDraft     OrderStatus = "DRAFT"
	StatusSubmitted OrderStatus = "SUBMITTED"
)

type LineItem struct {
	SKU string
	Qty int
}

type Order struct {
	ID     string
	status OrderStatus
	lines  []LineItem
}

func NewOrder(id string) *Order {
	return &Order{ID: id, status: StatusDraft}
}

func (o *Order) AddLine(item LineItem) error {
	if o.status != StatusDraft {
		return errors.New("cannot modify submitted order")
	}
	o.lines = append(o.lines, item)
	return nil
}

func (o *Order) Submit() error {
	if len(o.lines) == 0 {
		return errors.New("empty order")
	}
	o.status = StatusSubmitted
	return nil
}
